### PR TITLE
[JENKINS-68417] Revert `LogRecord` tricks

### PR DESCRIPTION
**Pull request by jglick**
See [JENKINS-68417](https://issues.jenkins.io/browse/JENKINS-68417).

Reverts #6018 and #6449; keeping synchronization simplifications from #6018, the correction to those in #6044, and (slightly amended) test coverage. ~All “published” log records (`INFO` and above, or finer if in a custom logger) are retained, but if there are parameters with a message format, that format is expanded at publishing time so that the parameters can be discarded, solving any memory leak issues at the cost of a bit of extra CPU work for custom loggers which the admin never actually looks at.~ https://github.com/jenkinsci/workflow-cps-plugin/pull/557 seems to suffice.

### Proposed changelog entries

* Reverting prior attempts to make log records collectible.

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/6643"></a>

### CODE REVIEW

File: core/src/main/java/hudson/util/RingBufferLogHandler.java
Comment (jglick) [inline core/src/main/java/hudson/util/RingBufferLogHandler.java:80]: I felt nervous about updating the `LogRecord` in place, though it would probably be thread-safe (since `log` call would consult `Handler`s serially) and would halve the work when combined with `support-core`.

File: core/src/main/java/hudson/util/RingBufferLogHandler.java
Comment (jglick) [inline core/src/main/java/hudson/util/RingBufferLogHandler.java:89]: Just to make sure any large objects that might be held by the stack trace are not also leaked.

File: core/src/main/java/hudson/util/RingBufferLogHandler.java
Comment (jglick) [inline core/src/main/java/hudson/util/RingBufferLogHandler.java:97]: hide whitespace in diff

File: core/src/main/java/hudson/util/RingBufferLogHandler.java
Comment (timja) [inline core/src/main/java/hudson/util/RingBufferLogHandler.java:97]: should this be using an actual logger?

Comment (jglick) [reply core/src/main/java/hudson/util/RingBufferLogHandler.java:97]: No, because that could cause a cycle of sorts. In general you have to be cautious about what you do in logger methods.

File: test/src/test/java/jenkins/model/JenkinsLogRecordsTest.java
Comment (basil) [inline test/src/test/java/jenkins/model/JenkinsLogRecordsTest.java:52]: The `ProxyException` logic is not covered by tests and is therefore subject to regression. Adding test coverage for this logic reduces this risk of regression:

```suggestion
            Logger logger = Logger.getLogger(Jenkins.class.getName());
            LogRecord lr = new LogRecord(Level.INFO, "formatting {0}");
            Object params[] = {x};
            lr.setLoggerName(logger.getName());
            lr.setParameters(params);
            lr.setThrown(new Throwable());
            logger.log(lr);
            WeakReference<Object> ref = new WeakReference<>(x);
            x = null;
            MemoryAssert.assertGC(ref, true);
            assertThat("Record parameters formatted before collection",
                logRecords.stream().map(LogRecord::getMessage).collect(Collectors.toList()),
                hasItem("formatting collect me"));
            assertThat("Large objects that might be held by the stack trace not held",
                 logRecords.stream()
                      .filter(record -> record.getMessage().contains("formatting collect me"))
                      .map(LogRecord::getThrown)
                      .findFirst()
                      .orElseThrow(() -> new AssertionError("not found")),
                 instanceOf(ProxyException.class));
```

(Plus corresponding imports.)

File: core/src/main/java/hudson/util/RingBufferLogHandler.java
Comment (basil) [inline core/src/main/java/hudson/util/RingBufferLogHandler.java:81]: A call to `super.isLoggable(record)` does this null check, and it also checks whether the `LogRecord` has an appropriate level and whether it satisfies any filter. In practice, the level check is handled by `hudson.logging.LogRecorder.Target#matches` and subclasses are unlikely to implement a filter, but calling `super.isLoggable(record)` here results in a slight deduplication of code and makes this subclass of `Handler` consistent with the subclasses of `Handler` provided by the Java Platform, so I think this is worth doing.

```suggestion
        if (!super.isLoggable(record)) {
            return;
        }
```

File: core/src/main/java/hudson/util/RingBufferLogHandler.java
Comment (basil) [inline core/src/main/java/hudson/util/RingBufferLogHandler.java:48]: There is already a field for a `Formatter` in the `Handler` class, which we could populate with an instance of `SimpleFormatter` in the constructor. Unlike the solution presented here, where a single `SimpleFormatter` is used for all `RingBufferLogHandler`s, using the field in the parent class would result in a separate `SimpleFormatter` for each `RingBufferLogHandler`. I think the solution offered here is more space-efficient at the cost of an additional field and an assumption that `Formatter#formatMessage` will always remain effectively static. I think it would be safer to avoid the latter assumption, even though the alternative is slightly less space-efficient, just in case `Formatter#formatMessage` is no longer effectively static in a future version of Java.

File: core/src/main/java/hudson/util/RingBufferLogHandler.java
Comment (basil) [inline core/src/main/java/hudson/util/RingBufferLogHandler.java:86]: Since `LogRecord#setMillis` is deprecated in Java 9+, and since we plan on requiring Java 11 very soon, I think it would be worthwhile to leave a comment reminding ourselves to update this usage in the future. Otherwise, we may very well forget to do this and continue using the deprecated method.

```suggestion
                clone.setMillis(record.getMillis()); // TODO Java 9+ clone.setInstance(record.getInstant());
```

File: core/src/main/java/hudson/util/RingBufferLogHandler.java
Comment (basil) [inline core/src/main/java/hudson/util/RingBufferLogHandler.java:82]: Good that we are not reintroducing the deadlocks experienced previously when trying to eagerly format log records on agent JVMs. My understanding is this PR eliminates `LogRecordRef`, which means that someone could now again (as they could before #6018) set up a custom logger that records messages from an agent JVM and whose parameter references may last beyond the end of the build. So in other words, this PR returns us to the pre-#6018 status quo with the addition of eager formatting on the controller side (but without eager formatting on the agent side, which carries additional implementation difficulties). This seems like a net improvement as long as the lack of agent-side support does not interfere with things like BOM/PCT testing (which was the original motivation behind working on this in #6018).

**Review by basil — APPROVED** [review]
Very nice! The original motivation behind #6018 was BOM/PCT failures, so I think we should run BOM/PCT on this to ensure that the partial revert of #6018 does not expose any old BOM/PCT failures again.

**Comment by jglick** [comment]
> we should run BOM/PCT on this

Yeah, would not hurt. Pretested the known relevant issue—see https://github.com/jenkinsci/support-core-plugin/pull/376.

File: test/src/test/java/jenkins/model/JenkinsLogRecordsTest.java
Comment (jglick) [reply test/src/test/java/jenkins/model/JenkinsLogRecordsTest.java:52]: Yes, it is probably wise to do this, though I would rather assert the _goal_ (that an object present in a field of the `Throwable` is not held) than the _mechanism_ (`ProxyException`).

File: core/src/main/java/hudson/util/RingBufferLogHandler.java
Comment (jglick) [reply core/src/main/java/hudson/util/RingBufferLogHandler.java:81]: I will have to double-check, but I think this is what we do _not_ want to do because this would prevent custom fine loggers from getting anything. Regular handlers either print a message to a sink or do not; handlers which collect all messages need to behave differently. https://github.com/jenkinsci/support-core-plugin/blob/7e75fdc7e9af0abe0d7d659a3e6e591c3451288f/src/main/java/com/cloudbees/jenkins/support/SupportLogHandler.java#L128-L135

File: core/src/main/java/hudson/util/RingBufferLogHandler.java
Comment (jglick) [reply core/src/main/java/hudson/util/RingBufferLogHandler.java:48]: The small number of handlers makes the space efficiency unimportant. I chose to use a separate field more for clarity: we are only calling `formatMessage` and _not_ `format` like the super field would normally be used for.

File: core/src/main/java/hudson/util/RingBufferLogHandler.java
Comment (jglick) [reply core/src/main/java/hudson/util/RingBufferLogHandler.java:86]: Indeed. Still on the fence about cloning, vs. just mutating the original `LogRecord`.

File: core/src/main/java/hudson/util/RingBufferLogHandler.java
Comment (jglick) [reply core/src/main/java/hudson/util/RingBufferLogHandler.java:82]: > we are not reintroducing the deadlocks experienced previously

That I know of. It is particularly difficult to reason about `Logger` behavior because it can be called from sensitive contexts such as class loading.

> the lack of agent-side support

Exactly, the assumption here is that controller JVMs are likely to last much longer than agent JVMs; a crash of an agent JVM is more easily recovered; and the known scenario motivating the original change (a giant `GroovyClassLoader` in a message format) was on the controller, not the agent.

Comment (basil) [reply core/src/main/java/hudson/util/RingBufferLogHandler.java:82]: > That I know of. It is particularly difficult to reason about Logger behavior because it can be called from sensitive contexts such as class loading.

I suppose there is a risk of deadlock when loading the `JenkinsJVM` class, just as there previously was with `LogRecordRef`. If that manifests itself in practice, I suppose we could solve that issue by preloading it as we did with `LogRecordRef`.

File: test/src/test/java/jenkins/model/JenkinsLogRecordsTest.java
Comment (basil) [reply test/src/test/java/jenkins/model/JenkinsLogRecordsTest.java:52]: Even a mere assertion of the mechanism prevents the mechanism from accidentally being undone by a future change, which is better than nothing. Obviously an assertion of the goal would be even better.

File: core/src/main/java/hudson/util/RingBufferLogHandler.java
Comment (basil) [reply core/src/main/java/hudson/util/RingBufferLogHandler.java:81]: > This would prevent custom fine loggers from getting anything

I do not see how it could because we never mutate `java.util.logging.Handler#logLevel` from its default value of `ALL` in `RingBufferLogHandler`. Like I wrote earlier, I do not think there is a difference in practice, but I think using the built-in method better adheres to the parent class's extensible design just in case a theoretical subclass of `RingBufferLogHandler` wanted to use a custom level or filter.

File: core/src/main/java/hudson/util/RingBufferLogHandler.java
Comment (basil) [reply core/src/main/java/hudson/util/RingBufferLogHandler.java:48]: Your logic for choosing to use a separate field makes sense to me, but I still think it is risky to assume that `formatMessage` is effectively static and thus that a single instance can be shared. That assumption may have been safe in the days of Java 8, but I feel that in recent years the Java standard libraries are not set in stone as much as they used to be.

**Comment by jglick** [comment]
My timebox on this issue is expiring. Considering throwing it all out (just keeping the reversion parts) and instead avoiding large record parameters on a case-by-case basis.

File: core/src/main/java/hudson/util/RingBufferLogHandler.java
Comment (jglick) [reply core/src/main/java/hudson/util/RingBufferLogHandler.java:72]: It appears to work. Still, I would rather not do it here because

* I am not entirely confident it is correct.
* It is not much related to the rest of the change. (I merely added the null check after noticing that `tooMuchRecordsShouldNotCrashHandler` was not testing anything.)

**Review by jtnord — APPROVED** [review]

File: core/src/main/java/hudson/util/RingBufferLogHandler.java
Comment (jglick) [reply core/src/main/java/hudson/util/RingBufferLogHandler.java]: > In general you have to be cautious

(advice I should have taken more to heart)

File: core/src/main/java/hudson/util/RingBufferLogHandler.java
Comment (basil) [reply core/src/main/java/hudson/util/RingBufferLogHandler.java:72]: > I would rather just make it final than consider such possibilities.

I am OK with making it `final`.

**Review by basil — APPROVED** [review]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

**Pull request merged**
